lib/saml/settings_service.rb

```diff
@@ -1,28 +1,43 @@
 # frozen_string_literal: true
 require 'memoist'
+require 'sentry_logging'
 
 module SAML
   # This class is responsible for putting together a complete ruby-saml
   # SETTINGS object, meaning, our static SP settings + the IDP settings
   # which must be fetched once and only once via IDP metadata.
   class SettingsService
     class << self
+      include SentryLogging
       extend Memoist
 
       METADATA_RETRIES = 3
       OPEN_TIMEOUT = 2
       TIMEOUT = 15
 
       def saml_settings
+        if @fetch_attempted.nil? || metadata_successfully_retrieved?
+          merged_saml_settings
+        else
+          # retrieve was attempted but failed, blow away memoized data and try again
+          merged_saml_settings(true)
```

Comment:
I understand now - you want to move the `use-cache-if-we-successfully-retrieved || GET-new-metadata-if-we-havent` logic over to the metadata method.

> This would be crappy if the call to OneLogin::RubySaml::IdpMetadataParser.new.parse is itself slow

The `metadata` in `parse(metadata, settings: settings` is ~10.5 kB of XML.  [`ruby-saml` does a number of `REXML` parsing operations on said `metadata` XML](https://github.com/onelogin/ruby-saml/blob/v1.3.0/lib/onelogin/ruby-saml/idp_metadata_parser.rb#L42) - so there _is_ performance gained in memoizing the RubySaml::Settings over only memoizing the raw `metadata` XML response.  I'm not exactly sure how much, but I don't think it's negligible.